### ISSUE TITLE
Aliasing generic Annotation can't specialize

### ISSUE THREAD

**Issue by JesseFarebro** [microsoft/pyright#6324]
## Environment data

-   Language Server version: v2023.11.10
-   OS and version: macOS 14.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10 (pyenv)

### `a.py`
```python
from typing import Annotated, TypeVar

T = TypeVar("T")

CustomAnnotation = Annotated[T, "custom_annotation"]
```

### `b.py`
```python
import a

CustomAnnotation = a.CustomAnnotation # This breaks the type

def f(x: CustomAnnotation[int]) -> str:
    return ""
```

## Repro Steps

1. If you define a generic annotated type in file `a.py` and you import the root module as `import a` and alias the type, e.g., `CustomAnnotation = a.CustomAnnotation`, this breaks.

## Expected behavior

No error, the annotation is valid.

## Actual behavior

```
Could not specialize type "type[T@CustomAnnotation]"
  Type "int" cannot be assigned to type "T@CustomAnnotation"Pylance
```

**Comment by erictraut** [comment]
@JesseFarebro, I'm closing as "won't fix" because I don't consider this a bug, but I'm happy to help you with a workaround if you'd like to provide more details about what you're trying to do here.

**Comment by hmc-cs-mdrissi** [comment]
There's one example in that [PEP](https://peps.python.org/pep-0593/#aliases-concerns-over-verbosity) that has an alias precisely like this one.

```python
T = TypeVar('T')
Const = Annotated[T, my_annotations.CONST]
```

I've occasionally run into this pattern as well. A way to attach some metadata to a type that I can re-use across types. The goal being I would want Const[int]/Const[str] to behave like int/str to pyright, but my own tooling/introspection can add behavior based on Const. I ended up just avoiding this pattern in the end as unsupported.

**Comment by erictraut** [comment]
@JesseFarebro, taking another look at your code, there's another reason it doesn't work. You've defined a generic type alias called `CustomAnnotation` in file `a.py`. It takes one type parameter. Then you define a second type alias in file `b.py` that's also called `CustomAnnotation`. The second type alias is not generic (because there are no type variables used in its definition). It's defined in terms of the first type alias, but you don't provide any type argument. That means it has an implied type argument of `Any`. It's equivalent to the following:

```python
CustomAnnotation = a.CustomAnnotation
```

is equivalent to:

```python
CustomAnnotation = a.CustomAnnotation[Any]
```

I think what you meant to do here is define another generic type alias, like the following:
```python
T = TypeVar("T")
CustomAnnotation = test15.CustomAnnotation[T]
```

If you do this, it type checks without errors.

A few additional tips that will help make your code more robust and allow you to more easily diagnose issues with type aliases:
* When defining a type alias, use the `TypeAlias` type annotation (see [PEP 613](https://peps.python.org/pep-0613/) for details). Or even better, if you're using Python 3.12, use the new dedicated `type` statement (see [PEP 695](https://peps.python.org/pep-0695/) for details).
* Enable pyright's `reportMissingTypeArgument` diagnostic check. This will emit an error if you forget to supply an explicit type argument for a generic class or type alias, as you've done in your code.

**Comment by JesseFarebro** [comment]
Wow, thanks, @erictraut, for the quick and detailed reply. This does make more sense now. A little context for this bug: I'm somewhat restrictive with module imports; I prefer to import a module instead of importing attributes of that module. With this in mind, I could easily do:

```py
import a

def f(x: a.CustomAnnotation[int]) -> str:
  ...
```

but sometimes, I prefer to redefine the alias in the file to save space / improve formatting. I'm not a massive fan of having to do

```py
T = TypeVar("T")
CustomAnnotation = a.CustomAnnotation[T]
```

but I understand why the alternative in my original post doesn't pass the type checker.

Also, thanks for the suggestion on `reportMissingTypeArgument` this will be useful in general 😄.
